// @telemetryos/cli/dist/services/api-client.js
import { loadCliConfig } from './cli-config.js';
const DEFAULT_API_URL = 'https://api.telemetryos.com';
export class ApiClient {
    constructor(baseUrl, token) {
        this.baseUrl = baseUrl;
        this.token = token;
    }
    static async create() {
        const config = await loadCliConfig();
        if (!config.apiToken) {
            throw new Error('Not authenticated. Run `tos auth` first.');
        }
        const baseUrl = process.env.TOS_API_URL || config.apiUrl || DEFAULT_API_URL;
        return new ApiClient(baseUrl, config.apiToken);
    }
    async get(path, opts) {
        return this.request('GET', path, opts);
    }
    async post(path, opts) {
        return this.request('POST', path, opts);
    }
    async put(path, opts) {
        return this.request('PUT', path, opts);
    }
    async patch(path, opts) {
        return this.request('PATCH', path, opts);
    }
    async delete(path, opts) {
        return this.request('DELETE', path, opts);
    }
    async request(method, path, opts) {
        const url = this.buildUrl(path, opts === null || opts === void 0 ? void 0 : opts.query);
        const headers = {
            Authorization: `Bearer ${this.token}`,
            ...opts === null || opts === void 0 ? void 0 : opts.headers,
        };
        // Don't set Content-Type for FormData - fetch sets it with boundary
        if ((opts === null || opts === void 0 ? void 0 : opts.body) && !(opts.body instanceof FormData)) {
            headers['Content-Type'] = 'application/json';
        }
        const response = await fetch(url, {
            method,
            headers,
            body: (opts === null || opts === void 0 ? void 0 : opts.body) instanceof FormData
                ? opts.body
                : (opts === null || opts === void 0 ? void 0 : opts.body)
                    ? JSON.stringify(opts.body)
                    : undefined,
        });
        if (!response.ok) {
            const errorBody = await response.text().catch(() => 'Unknown error');
            if (response.status === 401) {
                throw new Error('Authentication failed. Run `tos auth` to re-authenticate.');
            }
            throw new Error(`API Error ${response.status}: ${errorBody}`);
        }
        return response;
    }
    buildUrl(path, query) {
        const url = new URL(path, this.baseUrl);
        if (query) {
            for (const [key, value] of Object.entries(query)) {
                if (value !== undefined && value !== null) {
                    url.searchParams.append(key, String(value));
                }
            }
        }
        return url.toString();
    }
}
