// @telemetryos/cli/dist/services/archiver.js
import { readFile, stat, unlink } from 'fs/promises';
import path from 'path';
import { tmpdir } from 'os';
import * as tar from 'tar';
import ignoreModule from 'ignore';
const ignore = ignoreModule.default || ignoreModule;
const DEFAULT_IGNORES = [
    'node_modules',
    '.git',
    'dist',
    '.DS_Store',
    '*.log',
    '.env',
    '.env.*',
    '.turbo',
    'coverage',
];
export async function createArchive(projectPath, onProgress) {
    const projectName = path.basename(projectPath);
    const filename = `${projectName}-${Date.now()}.tar.gz`;
    const tempPath = path.join(tmpdir(), filename);
    // Build ignore filter
    const ig = ignore();
    // Load .gitignore patterns if present, otherwise use defaults
    try {
        const gitignoreContent = await readFile(path.join(projectPath, '.gitignore'), 'utf-8');
        ig.add(gitignoreContent);
        ig.add('.git'); // .git is always ignored by git itself
        onProgress === null || onProgress === void 0 ? void 0 : onProgress('Using .gitignore patterns');
    }
    catch {
        ig.add(DEFAULT_IGNORES);
        onProgress === null || onProgress === void 0 ? void 0 : onProgress('No .gitignore found, using default exclusions');
    }
    // Create tar.gz archive
    onProgress === null || onProgress === void 0 ? void 0 : onProgress('Creating archive...');
    await tar.create({
        gzip: true,
        file: tempPath,
        cwd: projectPath,
        filter: (filePath) => {
            // Always include root
            if (filePath === '.')
                return true;
            // Normalize path for ignore matching
            const relativePath = filePath.startsWith('./') ? filePath.slice(2) : filePath;
            const ignored = ig.ignores(relativePath);
            if (!ignored)
                onProgress === null || onProgress === void 0 ? void 0 : onProgress(`  ${relativePath}`);
            return !ignored;
        },
    }, ['.']);
    const buffer = await readFile(tempPath);
    const stats = await stat(tempPath);
    onProgress === null || onProgress === void 0 ? void 0 : onProgress(`Archive created: ${formatBytes(stats.size)}`);
    await unlink(tempPath).catch(() => { });
    return { buffer, filename };
}
function formatBytes(bytes) {
    if (bytes < 1024)
        return `${bytes} B`;
    if (bytes < 1024 * 1024)
        return `${(bytes / 1024).toFixed(1)} KB`;
    return `${(bytes / (1024 * 1024)).toFixed(1)} MB`;
}
