// @telemetryos/cli/dist/services/cli-config.js
import { mkdir, readFile, writeFile } from 'fs/promises';
import { homedir } from 'os';
import path from 'path';
import z from 'zod';
function getCliConfigDir() {
    const xdgConfig = process.env.XDG_CONFIG_HOME || path.join(homedir(), '.config');
    return path.join(xdgConfig, 'telemetryos', 'cli');
}
function getCliConfigPath() {
    return path.join(getCliConfigDir(), 'config.json');
}
export const apiTokenSchema = z
    .string()
    .regex(/^u[a-f0-9]{32}$/i, 'Token must start with "u" followed by 32 hex characters');
const cliConfigSchema = z.object({
    apiToken: z.string().optional(),
    apiUrl: z.url().optional(),
});
export async function loadCliConfig() {
    try {
        const configPath = getCliConfigPath();
        const content = await readFile(configPath, 'utf-8');
        const parsed = JSON.parse(content);
        return cliConfigSchema.parse(parsed);
    }
    catch (error) {
        if (error.code === 'ENOENT') {
            return cliConfigSchema.parse({});
        }
        const configPath = getCliConfigPath();
        if (error instanceof SyntaxError) {
            throw new Error(`Invalid JSON in CLI config file (${configPath}):\n  ${error.message}`);
        }
        if (error instanceof z.ZodError) {
            const issues = error.issues.map((i) => `  ${i.path.join('.')}: ${i.message}`).join('\n');
            throw new Error(`Invalid CLI config file (${configPath}):\n${issues}`);
        }
        throw error;
    }
}
export async function saveCliConfig(config) {
    const validated = cliConfigSchema.parse(config);
    await mkdir(getCliConfigDir(), { recursive: true });
    await writeFile(getCliConfigPath(), JSON.stringify(validated, null, 2));
}
