// @telemetryos/cli/dist/services/create-project.js
import { spawn, execSync } from 'child_process';
import fs from 'fs/promises';
import path from 'path';
import { ansi } from '../utils/ansi.js';
import { copyDir, templatesDir } from '../utils/template.js';
// Files that can exist in a directory without being considered conflicts
const safeExistingFiles = [
    '.DS_Store',
    '.git',
    '.gitignore',
    '.gitattributes',
    '.idea',
    '.vscode',
    'Thumbs.db',
    'LICENSE',
    'README.md',
];
export async function createProject(options) {
    const { name, description, author, version, template, claudeCode, projectPath, progressFn } = options;
    await fs.mkdir(projectPath, { recursive: true });
    // Initialize git repo early (before template dependencies that may have git hooks)
    const gitInitialized = tryGitInit(projectPath);
    if (gitInitialized) {
        console.log('\nInitialized a git repository');
    }
    await copyDir(path.join(templatesDir, template), projectPath, {
        name,
        version,
        description,
        author,
    }, progressFn);
    // Optionally apply Claude Code overlay (skills and settings)
    if (claudeCode) {
        await copyDir(path.join(templatesDir, 'claude-code'), projectPath, {
            name,
            version,
            description,
            author,
        }, progressFn);
    }
    // Install latest versions of @telemetryos/sdk and @telemetryos/cli
    console.log('\nInstalling dependencies...');
    await installPackages(projectPath);
    // Create initial commit after all files are in place
    if (gitInitialized) {
        await tryGitCommit(projectPath);
    }
    printSuccessMessage(name, projectPath);
}
async function installPackages(projectPath) {
    // Install SDK as a regular dependency
    await new Promise((resolve, reject) => {
        const sdkInstall = spawn('npm', ['install', '@telemetryos/sdk'], {
            cwd: projectPath,
            stdio: 'inherit',
            shell: true,
        });
        sdkInstall.on('close', (code) => {
            if (code === 0) {
                resolve();
            }
            else {
                reject(new Error(`npm install @telemetryos/sdk failed with code ${code}`));
            }
        });
        sdkInstall.on('error', (error) => {
            reject(error);
        });
    });
    // Install CLI as a dev dependency
    await new Promise((resolve, reject) => {
        const cliInstall = spawn('npm', ['install', '-D', '@telemetryos/cli'], {
            cwd: projectPath,
            stdio: 'inherit',
            shell: true,
        });
        cliInstall.on('close', (code) => {
            if (code === 0) {
                resolve();
            }
            else {
                reject(new Error(`npm install -D @telemetryos/cli failed with code ${code}`));
            }
        });
        cliInstall.on('error', (error) => {
            reject(error);
        });
    });
}
function isInGitRepository(cwd) {
    try {
        execSync('git rev-parse --is-inside-work-tree', { cwd, stdio: 'ignore' });
        return true;
    }
    catch {
        return false;
    }
}
function tryGitInit(projectPath) {
    try {
        execSync('git --version', { stdio: 'ignore' });
        if (isInGitRepository(projectPath)) {
            return false;
        }
        execSync('git init', { cwd: projectPath, stdio: 'ignore' });
        return true;
    }
    catch {
        return false;
    }
}
async function tryGitCommit(projectPath) {
    try {
        execSync('git add -A', { cwd: projectPath, stdio: 'ignore' });
        execSync('git commit -m "Initialize project using TelemetryOS CLI"', {
            cwd: projectPath,
            stdio: 'ignore',
        });
        console.log('Created initial commit');
        return true;
    }
    catch {
        // Commit failed (e.g., git user not configured)
        // Remove .git directory to avoid half-initialized state
        console.log('Git commit not created');
        console.log('Removing .git directory...');
        try {
            await fs.rm(path.join(projectPath, '.git'), { recursive: true, force: true });
        }
        catch {
            // Ignore cleanup errors
        }
        return false;
    }
}
export async function checkDirectoryConflicts(projectPath) {
    try {
        const entries = await fs.readdir(projectPath);
        return entries.filter((file) => !safeExistingFiles.includes(file));
    }
    catch (error) {
        // Directory doesn't exist, no conflicts
        if (error instanceof Error && 'code' in error && error.code === 'ENOENT') {
            return [];
        }
        throw error;
    }
}
export async function removeConflictingFiles(projectPath, conflicts) {
    for (const file of conflicts) {
        const filePath = path.join(projectPath, file);
        try {
            const stat = await fs.stat(filePath);
            if (stat.isDirectory()) {
                await fs.rm(filePath, { recursive: true, force: true });
            }
            else {
                await fs.unlink(filePath);
            }
        }
        catch (error) {
            // Ignore errors for files that may have already been removed
            console.error(`Warning: Could not remove ${file}: ${error instanceof Error ? error.message : error}`);
        }
    }
}
function printSuccessMessage(name, projectPath) {
    // Calculate relative path from cwd for cleaner display
    const relativePath = path.relative(process.cwd(), projectPath);
    const displayPath = relativePath || '.';
    console.log(`
${ansi.green}Success!${ansi.reset} Created ${ansi.bold}${name}${ansi.reset} at ${ansi.cyan}${displayPath}${ansi.reset}

Inside that directory, you can run:

  ${ansi.cyan}npm run dev${ansi.reset}
    Starts the development server

  ${ansi.cyan}npm run build${ansi.reset}
    Builds the app for production

You may begin with:

  ${ansi.cyan}cd ${displayPath}${ansi.reset}
  ${ansi.cyan}npm run dev${ansi.reset}

`);
}
