// @telemetryos/cli/dist/services/project-config.js
import { readFile } from 'fs/promises';
import path from 'path';
import z from 'zod';
const mountPointSchema = z.object({
    path: z.string(),
});
const backgroundWorkerSchema = z.object({
    path: z.string(),
});
const serverWorkerSchema = z.object({
    path: z.string(),
});
const containerSchema = z.object({
    image: z.string(),
});
export const projectConfigSchema = z.object({
    name: z.string().optional(),
    version: z.string().optional(),
    description: z.string().optional(),
    logoPath: z.string().optional(),
    thumbnailPath: z.string().optional(),
    mountPoints: z.record(z.string(), z.union([mountPointSchema, z.string()])).optional(),
    backgroundWorkers: z.record(z.string(), z.union([backgroundWorkerSchema, z.string()])).optional(),
    serverWorkers: z.record(z.string(), z.union([serverWorkerSchema, z.string()])).optional(),
    containers: z.record(z.string(), z.union([containerSchema, z.string()])).optional(),
    useSpaRouting: z.boolean().optional(),
    devServer: z
        .object({
        runCommand: z.string().optional(),
        url: z.string(),
    })
        .optional(),
});
export async function loadProjectConfig(projectPath) {
    const configPath = path.join(projectPath, 'telemetry.config.json');
    try {
        const content = await readFile(configPath, 'utf-8');
        const parsed = JSON.parse(content);
        return projectConfigSchema.parse(parsed);
    }
    catch (error) {
        if (error.code === 'ENOENT') {
            throw new Error(`No telemetry.config.json found in ${projectPath}`);
        }
        if (error instanceof z.ZodError) {
            const issues = error.issues.map((i) => `  ${i.path.join('.')}: ${i.message}`).join('\n');
            throw new Error(`Invalid telemetry.config.json:\n${issues}`);
        }
        if (error instanceof SyntaxError) {
            throw new Error(`Invalid JSON in telemetry.config.json: ${error.message}`);
        }
        throw error;
    }
}
