// @telemetryos/cli/dist/services/run-server.js
import { spawn } from 'child_process';
import { access, readFile } from 'fs/promises';
import { fileURLToPath } from 'url';
import http from 'http';
import path from 'path';
import { createInterface } from 'readline/promises';
import serveHandler from 'serve-handler';
import pkg from '../../package.json' with { type: 'json' };
import { loadProjectConfig } from './project-config.js';
import { ansi, ansiRegex } from '../utils/ansi.js';
// import { handlePublishCommand } from '../commands/publish.js'
const IMAGE_MIME_TYPES = {
    '.jpg': 'image/jpeg',
    '.jpeg': 'image/jpeg',
    '.png': 'image/png',
    '.gif': 'image/gif',
    '.svg': 'image/svg+xml',
    '.webp': 'image/webp',
};
export async function runServer(projectPath, flags) {
    printSplashScreen();
    projectPath = path.resolve(process.cwd(), projectPath);
    let projectConfig;
    try {
        projectConfig = await loadProjectConfig(projectPath);
    }
    catch (error) {
        console.error(error.message);
        process.exit(1);
    }
    await serveDevelopmentApplicationHostUI(projectPath, flags.port, projectConfig);
    await serveTelemetryApplication(projectPath, projectConfig);
    // Print ready message after Vite is confirmed ready
    printServerInfo(flags.port);
}
async function serveDevelopmentApplicationHostUI(projectPath, port, projectConfig) {
    const hostUiPath = await import.meta.resolve('@telemetryos/development-application-host-ui/dist');
    const serveConfig = { public: fileURLToPath(hostUiPath) };
    const server = http.createServer();
    server.on('request', async (req, res) => {
        var _a, _b;
        const url = new URL(req.url, `http://${req.headers.origin}`);
        if (url.pathname === '/__tos-config__') {
            res.setHeader('Content-Type', 'application/json');
            res.end(JSON.stringify(projectConfig));
            return;
        }
        if (url.pathname === '/__tos-logo__') {
            await serveImageFile(res, projectPath, projectConfig.logoPath, 'logo');
            return;
        }
        if (url.pathname === '/__tos-thumbnail__') {
            await serveImageFile(res, projectPath, projectConfig.thumbnailPath, 'thumbnail');
            return;
        }
        if (url.pathname === '/__dev_proxy__' && req.method === 'POST' && res) {
            let body = '';
            for await (const chunk of req) {
                body += chunk;
            }
            try {
                const { url, method, headers, body: requestBody } = JSON.parse(body);
                const response = await fetch(url, {
                    method,
                    headers,
                    body: requestBody !== null && requestBody !== void 0 ? requestBody : undefined,
                });
                const contentType = response.headers.get('content-type') || '';
                const isJson = contentType.includes('application/json');
                const isText = contentType.includes('text/') ||
                    contentType.includes('application/javascript') ||
                    contentType.includes('application/xml');
                let responseBody;
                let bodyType;
                if (isJson) {
                    const text = await response.text();
                    responseBody = text ? JSON.parse(text) : null;
                    bodyType = 'json';
                }
                else if (isText) {
                    responseBody = await response.text();
                    bodyType = 'text';
                }
                else {
                    // Binary data - convert to base64 for JSON transport
                    const arrayBuffer = await response.arrayBuffer();
                    responseBody = Buffer.from(arrayBuffer).toString('base64');
                    bodyType = 'binary';
                }
                const responseHeaders = {};
                response.headers.forEach((value, key) => {
                    responseHeaders[key] = value;
                });
                res.writeHead(200, { 'Content-Type': 'application/json' });
                res.end(JSON.stringify({
                    success: true,
                    status: response.status,
                    statusText: response.statusText,
                    headers: responseHeaders,
                    body: responseBody,
                    bodyType,
                    ok: response.ok,
                    url: response.url,
                }));
            }
            catch (error) {
                res.writeHead(200, { 'Content-Type': 'application/json' });
                res.end(JSON.stringify({
                    success: false,
                    errorMessage: `Proxy fetch failed: ${String(error)}`,
                    errorCause: error.cause ? String((_b = (_a = error.cause) === null || _a === void 0 ? void 0 : _a.message) !== null && _b !== void 0 ? _b : error.cause) : undefined,
                }));
            }
            return;
        }
        // TODO: Publish endpoint disabled until auth is working
        // if (url.pathname === '/__publish__') {
        //   if (req.method !== 'GET') {
        //     res.statusCode = 405
        //     res.end('Method not allowed')
        //     return
        //   }
        //
        //   // Set SSE headers
        //   res.setHeader('Content-Type', 'text/event-stream')
        //   res.setHeader('Cache-Control', 'no-cache')
        //   res.setHeader('Connection', 'keep-alive')
        //
        //   const sendEvent = (event: string, data: any) => {
        //     res.write(`event: ${event}\n`)
        //     res.write(`data: ${JSON.stringify(data)}\n\n`)
        //   }
        //
        //   try {
        //     sendEvent('state', { state: 'starting' })
        //
        //     await handlePublishCommand(
        //       projectPath,
        //       {},
        //       {
        //         onLog: (line: string) => {
        //           sendEvent('log', { message: line })
        //         },
        //         onStateChange: (state: string) => {
        //           sendEvent('state', { state })
        //         },
        //         onComplete: (data: { success: boolean; buildIndex?: number; duration?: string }) => {
        //           sendEvent('complete', data)
        //           res.end()
        //         },
        //         onError: (error: Error) => {
        //           const isAuthError =
        //             error.message.includes('authenticate') ||
        //             error.message.includes('Not authenticated')
        //           sendEvent('error', {
        //             error: error.message,
        //             code: isAuthError ? 'AUTH_REQUIRED' : 'PUBLISH_FAILED',
        //             message: isAuthError
        //               ? "Run 'tos auth' in the terminal to authenticate"
        //               : error.message,
        //           })
        //           res.end()
        //         },
        //       },
        //     )
        //   } catch (error) {
        //     sendEvent('error', {
        //       error: (error as Error).message,
        //       code: 'UNEXPECTED_ERROR',
        //       message: (error as Error).message,
        //     })
        //     res.end()
        //   }
        //   return
        // }
        serveHandler(req, res, serveConfig).catch((err) => {
            console.error('Error handling request:', err);
            res.statusCode = 500;
            res.end('Internal Server Error');
        });
    });
    server.listen(port);
}
async function serveImageFile(res, projectPath, filePath, label) {
    if (!filePath) {
        res.statusCode = 404;
        res.setHeader('Content-Type', 'text/plain; charset=utf-8');
        res.end(`No ${label} configured`);
        return;
    }
    const projectRoot = path.resolve(projectPath);
    const validatePath = (fullPath) => fullPath.startsWith(projectRoot + path.sep) || fullPath === projectRoot;
    // Try public/ first (Vite convention), then project root (backward compat)
    const publicPath = path.resolve(projectPath, 'public', filePath);
    const rootPath = path.resolve(projectPath, filePath);
    let resolvedPath;
    if (validatePath(publicPath)) {
        try {
            await access(publicPath);
            resolvedPath = publicPath;
        }
        catch { }
    }
    if (!resolvedPath && validatePath(rootPath)) {
        resolvedPath = rootPath;
    }
    if (!resolvedPath) {
        res.statusCode = 403;
        res.setHeader('Content-Type', 'text/plain; charset=utf-8');
        res.end('Forbidden: path escapes project root');
        return;
    }
    try {
        const imageData = await readFile(resolvedPath);
        const ext = path.extname(filePath).toLowerCase();
        const contentType = IMAGE_MIME_TYPES[ext] || 'application/octet-stream';
        res.setHeader('Content-Type', contentType);
        res.end(imageData);
    }
    catch (error) {
        res.setHeader('Content-Type', 'text/plain; charset=utf-8');
        if ((error === null || error === void 0 ? void 0 : error.code) === 'ENOENT') {
            res.statusCode = 404;
            res.end(`${label.charAt(0).toUpperCase() + label.slice(1)} file not found`);
        }
        else {
            res.statusCode = 500;
            res.end(`Error reading ${label}: ${(error === null || error === void 0 ? void 0 : error.message) || 'unknown error'}`);
        }
    }
}
async function serveTelemetryApplication(rootPath, projectConfig) {
    return new Promise((resolve, reject) => {
        var _a;
        if (!((_a = projectConfig === null || projectConfig === void 0 ? void 0 : projectConfig.devServer) === null || _a === void 0 ? void 0 : _a.runCommand)) {
            console.log('No value in config at devServer.runCommand');
            resolve();
            return;
        }
        const runCommand = projectConfig.devServer.runCommand;
        const binPath = path.join(rootPath, 'node_modules', '.bin');
        const childProcess = spawn(runCommand, {
            shell: true,
            env: {
                ...process.env,
                FORCE_COLOR: '1',
                PATH: `${binPath}${path.delimiter}${process.env.PATH}`,
            },
            stdio: ['ignore', 'pipe', 'pipe'],
            cwd: rootPath,
        });
        const stdoutReadline = createInterface({
            input: childProcess.stdout,
            crlfDelay: Infinity,
        });
        const stderrReadline = createInterface({
            input: childProcess.stderr,
            crlfDelay: Infinity,
        });
        let cleaned = false;
        const cleanup = () => {
            if (cleaned)
                return;
            cleaned = true;
            process.removeListener('exit', onParentExit);
            clearTimeout(timeoutHandle);
            childProcess.kill();
            stdoutReadline.close();
            stderrReadline.close();
        };
        const onParentExit = () => {
            console.log('Shutting down development server...');
            cleanup();
        };
        process.on('exit', onParentExit);
        const timeoutHandle = setTimeout(() => {
            cleanup();
            reject(new Error('Vite dev server did not start within 30 seconds'));
        }, 30000);
        let viteReady = false;
        stdoutReadline.on('line', (line) => {
            console.log(`[application]: ${line}`);
            // Detect Vite ready signal
            if (!viteReady) {
                const cleanLine = line.replace(ansiRegex, '');
                if (cleanLine.includes('VITE') && cleanLine.includes('ready in')) {
                    viteReady = true;
                    clearTimeout(timeoutHandle);
                    resolve();
                }
            }
        });
        stderrReadline.on('line', (line) => {
            console.error(`[application]: ${line}`);
        });
        childProcess.on('error', (error) => {
            if (!viteReady) {
                cleanup();
                reject(error);
            }
        });
        childProcess.on('close', (code) => {
            if (!viteReady) {
                cleanup();
                reject(new Error(`Dev server process exited with code ${code} before becoming ready`));
            }
        });
    });
}
function printSplashScreen() {
    console.log(`

     ${ansi.white} █       █                  █          ${ansi.yellow}▄▀▀▀▄ ▄▀▀▀▄
     ${ansi.white} █       █                  █          ${ansi.yellow}█   █ █
     ${ansi.white}▀█▀ ▄▀▀▄ █ ▄▀▀▄ █▀▄▀▄ ▄▀▀▄ ▀█▀ █▄▀ █ █ ${ansi.yellow}█   █  ▀▀▀▄
     ${ansi.white} █  █▀▀▀ █ █▀▀▀ █ █ █ █▀▀▀  █  █   █ █ ${ansi.yellow}█   █     █
     ${ansi.white} ▀▄ ▀▄▄▀ █ ▀▄▄▀ █ █ █ ▀▄▄▀  ▀▄ █    █  ${ansi.yellow}▀▄▄▄▀ ▀▄▄▄▀
     ${ansi.white}                                   ▄▀  ${ansi.reset}
     v${pkg.version}`);
}
function printServerInfo(port) {
    console.log(`
╔═══════════════════════════════════════════════════════════╗
║ ${ansi.bold}Development environment running at: ${ansi.cyan}http://localhost:${port}${ansi.reset} ║
╚═══════════════════════════════════════════════════════════╝
`);
}
