// @telemetryos/sdk/dist/react.cjs
"use strict";Object.defineProperty(exports,Symbol.toStringTag,{value:"Module"});const u=require("react"),x=require("./index.cjs");var O={exports:{}},j={};/**
 * @license React
 * react-jsx-runtime.production.js
 *
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */var q;function oe(){if(q)return j;q=1;var n=Symbol.for("react.transitional.element"),s=Symbol.for("react.fragment");function i(o,a,c){var p=null;if(c!==void 0&&(p=""+c),a.key!==void 0&&(p=""+a.key),"key"in a){c={};for(var f in a)f!=="key"&&(c[f]=a[f])}else c=a;return a=c.ref,{$$typeof:n,type:o,key:p,ref:a!==void 0?a:null,props:c}}return j.Fragment=s,j.jsx=i,j.jsxs=i,j}var R={};/**
 * @license React
 * react-jsx-runtime.development.js
 *
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */var B;function ie(){return B||(B=1,process.env.NODE_ENV!=="production"&&(function(){function n(e){if(e==null)return null;if(typeof e=="function")return e.$$typeof===te?null:e.displayName||e.name||null;if(typeof e=="string")return e;switch(e){case E:return"Fragment";case N:return"Profiler";case _:return"StrictMode";case Z:return"Suspense";case K:return"SuspenseList";case ne:return"Activity"}if(typeof e=="object")switch(typeof e.tag=="number"&&console.error("Received an unexpected object in getComponentNameFromType(). This is likely a bug in React. Please file an issue."),e.$$typeof){case P:return"Portal";case X:return e.displayName||"Context";case C:return(e._context.displayName||"Context")+".Consumer";case Q:var r=e.render;return e=e.displayName,e||(e=r.displayName||r.name||"",e=e!==""?"ForwardRef("+e+")":"ForwardRef"),e;case ee:return r=e.displayName||null,r!==null?r:n(e.type)||"Memo";case F:r=e._payload,e=e._init;try{return n(e(r))}catch{}}return null}function s(e){return""+e}function i(e){try{s(e);var r=!1}catch{r=!0}if(r){r=console;var l=r.error,d=typeof Symbol=="function"&&Symbol.toStringTag&&e[Symbol.toStringTag]||e.constructor.name||"Object";return l.call(r,"The provided key is an unsupported type %s. This value must be coerced to a string before using it here.",d),s(e)}}function o(e){if(e===E)return"<>";if(typeof e=="object"&&e!==null&&e.$$typeof===F)return"<...>";try{var r=n(e);return r?"<"+r+">":"<...>"}catch{return"<...>"}}function a(){var e=L.A;return e===null?null:e.getOwner()}function c(){return Error("react-stack-top-frame")}function p(e){if($.call(e,"key")){var r=Object.getOwnPropertyDescriptor(e,"key").get;if(r&&r.isReactWarning)return!1}return e.key!==void 0}function f(e,r){function l(){D||(D=!0,console.error("%s: `key` is not a prop. Trying to access it will result in `undefined` being returned. If you need to access the same value within the child component, you should pass it as a different prop. (https://react.dev/link/special-props)",r))}l.isReactWarning=!0,Object.defineProperty(e,"key",{get:l,configurable:!0})}function k(){var e=n(this.type);return Y[e]||(Y[e]=!0,console.error("Accessing element.ref was removed in React 19. ref is now a regular prop. It will be removed from the JSX Element type in a future release.")),e=this.props.ref,e!==void 0?e:null}function m(e,r,l,d,A,U){var g=l.ref;return e={$$typeof:T,type:e,key:r,props:l,_owner:d},(g!==void 0?g:null)!==null?Object.defineProperty(e,"ref",{enumerable:!1,get:k}):Object.defineProperty(e,"ref",{enumerable:!1,value:null}),e._store={},Object.defineProperty(e._store,"validated",{configurable:!1,enumerable:!1,writable:!0,value:0}),Object.defineProperty(e,"_debugInfo",{configurable:!1,enumerable:!1,writable:!0,value:null}),Object.defineProperty(e,"_debugStack",{configurable:!1,enumerable:!1,writable:!0,value:A}),Object.defineProperty(e,"_debugTask",{configurable:!1,enumerable:!1,writable:!0,value:U}),Object.freeze&&(Object.freeze(e.props),Object.freeze(e)),e}function w(e,r,l,d,A,U){var g=r.children;if(g!==void 0)if(d)if(re(g)){for(d=0;d<g.length;d++)h(g[d]);Object.freeze&&Object.freeze(g)}else console.error("React.jsx: Static children should always be an array. You are likely explicitly calling React.jsxs or React.jsxDEV. Use the Babel transform instead.");else h(g);if($.call(r,"key")){g=n(e);var y=Object.keys(r).filter(function(se){return se!=="key"});d=0<y.length?"{key: someKey, "+y.join(": ..., ")+": ...}":"{key: someKey}",J[g+d]||(y=0<y.length?"{"+y.join(": ..., ")+": ...}":"{}",console.error(`A props object containing a "key" prop is being spread into JSX:
  let props = %s;
  <%s {...props} />
React keys must be passed directly to JSX without using spread:
  let props = %s;
  <%s key={someKey} {...props} />`,d,g,y,g),J[g+d]=!0)}if(g=null,l!==void 0&&(i(l),g=""+l),p(r)&&(i(r.key),g=""+r.key),"key"in r){l={};for(var z in r)z!=="key"&&(l[z]=r[z])}else l=r;return g&&f(l,typeof e=="function"?e.displayName||e.name||"Unknown":e),m(e,g,l,a(),A,U)}function h(e){v(e)?e._store&&(e._store.validated=1):typeof e=="object"&&e!==null&&e.$$typeof===F&&(e._payload.status==="fulfilled"?v(e._payload.value)&&e._payload.value._store&&(e._payload.value._store.validated=1):e._store&&(e._store.validated=1))}function v(e){return typeof e=="object"&&e!==null&&e.$$typeof===T}var S=u,T=Symbol.for("react.transitional.element"),P=Symbol.for("react.portal"),E=Symbol.for("react.fragment"),_=Symbol.for("react.strict_mode"),N=Symbol.for("react.profiler"),C=Symbol.for("react.consumer"),X=Symbol.for("react.context"),Q=Symbol.for("react.forward_ref"),Z=Symbol.for("react.suspense"),K=Symbol.for("react.suspense_list"),ee=Symbol.for("react.memo"),F=Symbol.for("react.lazy"),ne=Symbol.for("react.activity"),te=Symbol.for("react.client.reference"),L=S.__CLIENT_INTERNALS_DO_NOT_USE_OR_WARN_USERS_THEY_CANNOT_UPGRADE,$=Object.prototype.hasOwnProperty,re=Array.isArray,I=console.createTask?console.createTask:function(){return null};S={react_stack_bottom_frame:function(e){return e()}};var D,Y={},W=S.react_stack_bottom_frame.bind(S,c)(),V=I(o(c)),J={};R.Fragment=E,R.jsx=function(e,r,l){var d=1e4>L.recentlyCreatedOwnerStacks++;return w(e,r,l,!1,d?Error("react-stack-top-frame"):W,d?I(o(e)):V)},R.jsxs=function(e,r,l){var d=1e4>L.recentlyCreatedOwnerStacks++;return w(e,r,l,!0,d?Error("react-stack-top-frame"):W,d?I(o(e)):V)}})()),R}var H;function ae(){return H||(H=1,process.env.NODE_ENV==="production"?O.exports=oe():O.exports=ie()),O.exports}var t=ae();const ce=`:root {
  --background: 0 0% 100%;
  --foreground: 222.2 84% 4.9%;

  --muted: 220 12% 95%;
  --muted-foreground: 215.4 16.3% 46.9%;

  --border: 214.3 31.8% 91.4%;
  --input: 214.3 31.8% 91.4%;

  --card: 0 0% 100%;
  --card-foreground: 222.2 84% 4.9%;

  --accent: 220 12% 95%;
  --accent-foreground: 222.2 47.4% 11.2%;

  --primary: 39 93% 59%;
  --primary-foreground: 210 40% 98%;

  --secondary: 210 40% 96.1%;
  --secondary-foreground: 222.2 47.4% 11.2%;

  --destructive: 6 86% 61%;
  --destructive-foreground: 210 40% 98%;

  --ring: 222.2 84% 4.9%;

  --radius: 0.5rem;
}

.dark {
  --background: 210 28% 8%;
  --foreground: 210 40% 88%;

  --muted: 212, 28%, 10%;
  --muted-foreground: 215 20.2% 65.1%;

  --border: 212 24% 19%;
  --input: 210 5% 31%;

  --card: 212 28% 10%;
  --card-foreground: 210 40% 98%;

  --accent: 211 30% 15%;
  --accent-foreground: 210 40% 98%;

  --primary: 39 93% 59%;
  --primary-foreground: 210 40% 98%;

  --secondary: 217.2 32.6% 17.5%;
  --secondary-foreground: 210 40% 98%;

  --destructive: 6 86% 61%;
  --destructive-foreground: 210 40% 98%;

  --ring: 212.7 26.8% 83.9%;

  --radius: 0.5rem;
}

/* =============================================================================
   Container & Layout
   ============================================================================= */

.settings__container {
  padding: 16px;
  background: hsl(var(--background));
  color: hsl(var(--foreground));
  font-family: 'Rubik', sans-serif;
  font-feature-settings: "rlig" 1, "calt" 1;
  flex-grow: 1;
  display: flex;
  flex-direction: column;
  gap: 16px;
}

.settings__container>form {
  display: contents;
}

.settings__box {
  border: 1px solid hsl(var(--border));
  border-radius: var(--radius);
  padding: 16px;
  display: flex;
  flex-direction: column;
  gap: 16px;
}

.settings__heading {
  font-size: 16px;
  font-weight: bold;
  color: hsl(var(--foreground));
}

.settings__divider {
  border: solid hsl(var(--border));
  border-width: 1px 0 0;
  width: 100%;
}

/* =============================================================================
   Field Structure
   ============================================================================= */

.settings__field {
  display: inline-block;
}

.settings__label {
  font-size: 14px;
  margin-bottom: 6px;
  color: hsl(var(--foreground));
}

.settings__hint {
  font-size: 12px;
  opacity: 0.65;
  margin-top: 4px;
  color: hsl(var(--foreground));
}

.settings__error {
  font-size: 12px;
  margin-top: 4px;
  color: #ff6b6b;
}

/* =============================================================================
   Text Inputs
   ============================================================================= */

.settings__input-frame>* {
  height: 40px;
  width: 100%;
  font-size: 14px;
  padding: 8px 12px;
  border: 1px solid hsl(var(--border));
  border-radius: var(--radius);
  background: hsl(var(--background));
  color: hsl(var(--foreground));
}

.settings__input-frame>*:focus {
  outline: 1px solid hsl(var(--primary));
  outline-offset: 0;
}

.settings__input-frame>*::placeholder {
  color: hsl(var(--muted-foreground));
  opacity: 1;
}

.settings__textarea-frame>textarea {
  width: 100%;
  min-height: 80px;
  font-family: inherit;
  font-size: 14px;
  line-height: 1.4em;
  padding: 8px 12px;
  border: 1px solid hsl(var(--border));
  border-radius: var(--radius);
  background: hsl(var(--background));
  color: hsl(var(--foreground));
  resize: vertical;
}

.settings__textarea-frame>textarea:focus {
  outline: 1px solid hsl(var(--primary));
  outline-offset: 0;
}

.settings__textarea-frame>textarea::placeholder {
  color: hsl(var(--muted-foreground));
  opacity: 1;
}

/* =============================================================================
   Selection Inputs
   ============================================================================= */

.settings__select-frame>select {
  appearance: none;
  height: 40px;
  width: 100%;
  font-size: 14px;
  padding: 8px 36px 8px 12px;
  border: 1px solid hsl(var(--border));
  border-radius: var(--radius);
  background-color: hsl(var(--background));
  background-image: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='16' height='16' viewBox='0 0 24 24' fill='none' stroke='%239ca3af' stroke-width='2' stroke-linecap='round' stroke-linejoin='round'%3E%3Cpath d='m6 9 6 6 6-6'/%3E%3C/svg%3E");
  background-repeat: no-repeat;
  background-position: right 12px center;
  color: hsl(var(--foreground));
  cursor: pointer;
}

.settings__select-frame>select:hover {
  background-color: hsl(var(--accent));
}

.settings__select-frame>select:focus {
  outline: 1px solid hsl(var(--primary));
  outline-offset: 0;
}

.settings__slider-frame {
  display: flex;
  align-items: center;
  gap: 8px;
}

.settings__slider-frame>input[type="range"] {
  flex: 1;
  height: 8px;
  background: hsl(var(--secondary));
  border-radius: 9999px;
  outline: none;
  -webkit-appearance: none;
  appearance: none;
}

.settings__slider-frame>input[type="range"]:focus,
.settings__slider-frame>input[type="range"]:active {
  background: hsl(var(--primary));
}

.settings__slider-frame>input[type="range"]::-webkit-slider-thumb {
  -webkit-appearance: none;
  appearance: none;
  width: 20px;
  height: 20px;
  background: hsl(var(--background));
  border: 2px solid hsl(var(--primary));
  border-radius: 50%;
  cursor: pointer;
}

.settings__slider-frame>input[type="range"]::-moz-range-thumb {
  width: 20px;
  height: 20px;
  background: hsl(var(--background));
  border: 2px solid hsl(var(--primary));
  border-radius: 50%;
  cursor: pointer;
}

.settings__slider-frame>span {
  min-width: 36px;
  text-align: right;
  font-size: 14px;
  color: hsl(var(--muted-foreground));
}

.settings__slider-ruler {
  display: flex;
  justify-content: space-between;
  font-size: 12px;
  margin-top: 4px;
  margin-left: -4px;
  margin-right: 36px;
  opacity: 0.65;
}

.settings__color-frame {
  display: flex;
  align-items: center;
  gap: 8px;
}

.settings__color-frame>input[type="color"] {
  appearance: none;
  width: 40px;
  height: 40px;
  border: 1px solid hsl(var(--border));
  border-radius: var(--radius);
  padding: 0;
  cursor: pointer;
  flex-shrink: 0;
}

.settings__color-frame>input[type="color"]::-webkit-color-swatch-wrapper {
  padding: 2px;
}

.settings__color-frame>input[type="color"]::-webkit-color-swatch {
  border: none;
  border-radius: calc(var(--radius) - 3px);
}

.settings__color-frame>span {
  font-size: 14px;
  color: hsl(var(--foreground));
  font-family: monospace;
}

/* =============================================================================
   Toggle Inputs
   ============================================================================= */

.settings__switch-frame {
  display: flex;
  align-items: center;
  gap: 8px;
  font-size: 14px;
  cursor: pointer;
}

.settings__switch-frame>input[type="checkbox"] {
  appearance: none;
  width: 44px;
  height: 24px;
  background: hsl(var(--input));
  border-radius: 12px;
  position: relative;
  cursor: pointer;
  transition: background 0.2s;
  flex-shrink: 0;
}

.settings__switch-frame>input[type="checkbox"]::after {
  content: '';
  position: absolute;
  top: 2px;
  left: 2px;
  width: 20px;
  height: 20px;
  background: hsl(var(--background));
  border-radius: 50%;
  box-shadow: 0 1px 3px rgba(0, 0, 0, 0.3);
  transition: transform 0.2s;
}

.settings__switch-frame>input[type="checkbox"]:checked {
  background: hsl(var(--primary));
}

.settings__switch-frame>input[type="checkbox"]:checked::after {
  transform: translateX(20px);
}

.settings__switch-label {
  flex: 1;
  color: hsl(var(--foreground));
}

.settings__checkbox-frame {
  display: flex;
  align-items: center;
  gap: 8px;
  font-size: 14px;
  cursor: pointer;
}

.settings__checkbox-frame>input[type="checkbox"] {
  appearance: none;
  width: 16px;
  height: 16px;
  background: hsl(var(--background));
  border: 1px solid hsl(var(--primary));
  border-radius: 2px;
  cursor: pointer;
  position: relative;
}

.settings__checkbox-frame>input[type="checkbox"]:checked {
  background: hsl(var(--primary));
}

.settings__checkbox-frame>input[type="checkbox"]:checked::after {
  content: '';
  position: absolute;
  top: 1px;
  left: 5px;
  width: 4px;
  height: 8px;
  border: solid hsl(var(--primary-foreground));
  border-width: 0 2px 2px 0;
  transform: rotate(45deg);
}

.settings__checkbox-label {
  flex: 1;
  color: hsl(var(--foreground));
}

.settings__radio-frame {
  display: flex;
  align-items: center;
  gap: 8px;
  font-size: 14px;
  cursor: pointer;
}

.settings__radio-frame>input[type="radio"] {
  appearance: none;
  width: 16px;
  height: 16px;
  background: hsl(var(--background));
  border: 1px solid hsl(var(--primary));
  border-radius: 50%;
  cursor: pointer;
  position: relative;
}

.settings__radio-frame>input[type="radio"]:checked {
  background: hsl(var(--primary));
}

.settings__radio-frame>input[type="radio"]:checked::after {
  content: '';
  position: absolute;
  top: 4px;
  left: 4px;
  width: 6px;
  height: 6px;
  background: hsl(var(--primary-foreground));
  border-radius: 50%;
}

.settings__radio-label {
  flex: 1;
  color: hsl(var(--foreground));
}

/* =============================================================================
   Actions
   ============================================================================= */

.settings__button-frame {
  display: contents;
}

.settings__button-frame>button {
  padding: 8px 16px;
  background: hsl(var(--background));
  border: 1px solid hsl(var(--border));
  border-radius: var(--radius);
  color: hsl(var(--foreground));
  font-size: 14px;
  font-family: inherit;
  cursor: pointer;
}

.settings__button-frame>button:hover {
  background: hsl(var(--accent));
}

.settings__button-frame>button.active,
.settings__button-frame>button[data-active="true"] {
  background: hsl(var(--primary));
  border-color: hsl(var(--primary));
  color: hsl(var(--primary-foreground));
}

/* =============================================================================
   Media Select
   ============================================================================= */

.settings__media-select {
  position: relative;
  display: inline-block;
}

.settings__media-select-trigger {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 120px;
  height: 120px;
  padding: 0;
  border: 1px solid hsl(var(--border));
  border-radius: var(--radius);
  background: hsl(var(--background));
  color: hsl(var(--foreground));
  cursor: pointer;
  overflow: hidden;
}

.settings__media-select-trigger:hover {
  border-color: hsl(var(--primary));
}

.settings__media-select-trigger:focus {
  outline: 2px solid hsl(var(--primary));
  outline-offset: 2px;
}

.settings__media-select-trigger:disabled {
  opacity: 0.5;
  cursor: not-allowed;
}

.settings__media-select-trigger:disabled:hover {
  border-color: hsl(var(--border));
}

.settings__media-select-thumbnail {
  width: 100%;
  height: 100%;
  object-fit: cover;
}

.settings__media-select-thumbnail--empty {
  width: 100%;
  height: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
  background: hsl(var(--muted));
  color: hsl(var(--muted-foreground));
}

.settings__media-select-thumbnail--empty svg {
  width: 32px;
  height: 32px;
}

.settings__media-select-clear {
  position: absolute;
  top: -8px;
  right: -8px;
  width: 24px;
  height: 24px;
  padding: 0;
  background: hsl(var(--destructive));
  border: 2px solid hsl(var(--background));
  border-radius: 50%;
  cursor: pointer;
  color: hsl(var(--destructive-foreground));
  display: flex;
  align-items: center;
  justify-content: center;
  transition: transform 0.15s ease;
}

.settings__media-select-clear:hover {
  transform: scale(1.1);
}

.settings__media-select-clear svg {
  width: 12px;
  height: 12px;
}

/* Note: Modal styles are defined in Studio-UI where the picker modal is rendered */

/* =============================================================================
   Sections (collapsible accordion)
   ============================================================================= */

.settings__section {
  border: 1px solid hsl(var(--border));
  border-radius: var(--radius);
  overflow: hidden;
}

.settings__section-header {
  display: flex;
  align-items: center;
  width: 100%;
  padding: 12px 16px;
  background: hsl(var(--card));
  border: none;
  color: hsl(var(--foreground));
  font-size: 14px;
  font-weight: 600;
  font-family: inherit;
  cursor: pointer;
  gap: 8px;
}

.settings__section-header:hover {
  background: hsl(var(--accent));
}

.settings__section-arrow {
  display: inline-block;
  font-size: 10px;
  color: hsl(var(--muted-foreground));
  transition: transform 0.2s ease;
  transform: rotate(0deg);
}

.settings__section-arrow--open {
  transform: rotate(90deg);
}

.settings__section-title {
  flex: 1;
  text-align: left;
}

.settings__section-body {
  display: grid;
  grid-template-rows: 0fr;
  transition: grid-template-rows 0.2s ease;
}

.settings__section-body--open {
  grid-template-rows: 1fr;
}

.settings__section-content {
  overflow: hidden;
  min-height: 0;
  display: flex;
  flex-direction: column;
  gap: 16px;
  padding: 0 16px;
  transition: padding 0.2s ease;
}

.settings__section-body--open .settings__section-content {
  padding: 16px;
}
`;let G=!1;function M(){if(G||typeof document>"u")return;const n=document.createElement("style");n.setAttribute("data-telemetryos-sdk",""),n.textContent=ce,document.head.appendChild(n),G=!0}M();function le({children:n}){const s=()=>{const i=window.matchMedia("(prefers-color-scheme: dark)"),o=p=>{let f;p==="system"?f=i.matches:f=p!=="light",f?document.documentElement.classList.add("dark"):document.documentElement.classList.remove("dark")},a=x.environment().subscribeColorScheme(o);a.catch(console.error);const c=()=>{x.environment().getColorScheme().then(o).catch(console.error)};return i.addEventListener("change",c),()=>{a.then(()=>x.environment().unsubscribeColorScheme(o)).catch(console.error),i.removeEventListener("change",c)}};return u.useEffect(s,[]),t.jsx("div",{className:"settings__container",children:n})}function ue({children:n}){return t.jsx("div",{className:"settings__box",children:n})}function de({children:n}){return t.jsx("div",{className:"settings__heading",children:n})}function ge(){return t.jsx("hr",{className:"settings__divider"})}function fe({children:n}){return t.jsx("label",{className:"settings__field",children:n})}function pe({children:n}){return t.jsx("div",{className:"settings__label",children:n})}function me({children:n}){return t.jsx("div",{className:"settings__hint",children:n})}function he({children:n}){return t.jsx("div",{className:"settings__error",children:n})}function _e({children:n}){return t.jsx("div",{className:"settings__input-frame",children:n})}function xe({children:n}){return t.jsx("div",{className:"settings__textarea-frame",children:n})}function be({children:n}){return t.jsx("div",{className:"settings__select-frame",children:n})}function ve({children:n}){return t.jsx("div",{className:"settings__slider-frame",children:n})}function Se({children:n}){return t.jsx("div",{className:"settings__slider-ruler",children:n})}function ye({children:n}){return t.jsx("div",{className:"settings__color-frame",children:n})}function ke({children:n}){return t.jsx("label",{className:"settings__switch-frame",children:n})}function we({children:n}){return t.jsx("span",{className:"settings__switch-label",children:n})}function Ee({children:n}){return t.jsx("label",{className:"settings__checkbox-frame",children:n})}function je({children:n}){return t.jsx("span",{className:"settings__checkbox-label",children:n})}function Re({children:n}){return t.jsx("label",{className:"settings__radio-frame",children:n})}function Te({children:n}){return t.jsx("span",{className:"settings__radio-label",children:n})}function Ne({children:n}){return t.jsx("div",{className:"settings__button-frame",children:n})}M();function Ce({value:n,onChange:s,disabled:i=!1,placeholder:o="Select media...",accept:a}){const[c,p]=u.useState(!1),f=async()=>{if(!(i||c)){p(!0);try{const m=await x.media().openPicker({accept:a,currentValue:n??void 0});m&&s(m)}catch(m){console.error("Failed to open media picker:",m)}finally{p(!1)}}},k=m=>{m.stopPropagation(),s(null)};return t.jsxs("div",{className:"settings__media-select",children:[t.jsx("button",{type:"button",className:`settings__media-select-trigger ${n?"settings__media-select-trigger--has-value":""}`,onClick:f,disabled:i||c,title:n?`${n.name} - Click to change`:o,children:n?t.jsx("img",{src:n.thumbnailUrl,alt:n.name,className:"settings__media-select-thumbnail"}):t.jsx("div",{className:"settings__media-select-thumbnail settings__media-select-thumbnail--empty",children:c?t.jsx(Pe,{}):t.jsx(Oe,{})})}),n&&!i&&t.jsx("button",{type:"button",className:"settings__media-select-clear",onClick:k,title:"Remove image",children:t.jsx(Ae,{})})]})}function Ae(){return t.jsxs("svg",{width:"16",height:"16",viewBox:"0 0 24 24",fill:"none",stroke:"currentColor",strokeWidth:"2",children:[t.jsx("line",{x1:"18",y1:"6",x2:"6",y2:"18"}),t.jsx("line",{x1:"6",y1:"6",x2:"18",y2:"18"})]})}function Oe(){return t.jsxs("svg",{width:"16",height:"16",viewBox:"0 0 24 24",fill:"none",stroke:"currentColor",strokeWidth:"2",children:[t.jsx("rect",{x:"3",y:"3",width:"18",height:"18",rx:"2",ry:"2"}),t.jsx("circle",{cx:"8.5",cy:"8.5",r:"1.5"}),t.jsx("polyline",{points:"21 15 16 10 5 21"})]})}function Pe(){return t.jsxs("svg",{width:"16",height:"16",viewBox:"0 0 24 24",fill:"none",stroke:"currentColor",strokeWidth:"2",children:[t.jsx("circle",{cx:"12",cy:"12",r:"10",opacity:"0.25"}),t.jsx("path",{d:"M12 2a10 10 0 0 1 10 10",strokeLinecap:"round",children:t.jsx("animateTransform",{attributeName:"transform",type:"rotate",from:"0 12 12",to:"360 12 12",dur:"1s",repeatCount:"indefinite"})})]})}M();function Fe({title:n,defaultOpen:s=!0,children:i}){const[o,a]=u.useState(s),c=u.useId();return t.jsxs("div",{className:"settings__section",children:[t.jsxs("button",{type:"button",className:"settings__section-header",onClick:()=>a(!o),"aria-expanded":o,"aria-controls":c,children:[t.jsx("span",{className:`settings__section-arrow ${o?"settings__section-arrow--open":""}`,children:"▶"}),t.jsx("span",{className:"settings__section-title",children:n})]}),t.jsx("div",{id:c,className:`settings__section-body ${o?"settings__section-body--open":""}`,...!o&&{inert:!0},children:t.jsx("div",{className:"settings__section-content",children:i})})]})}function b(n,s,i,o=0){var a;const[c,p]=u.useState(!0),f=u.useRef(!0),[k]=u.useState(i),[m,w]=u.useState(void 0),[h,v]=u.useState(void 0),S=u.useCallback(_=>{f.current?console.error("Cannot set store value while loading. Destructure and check the first element of the returned tuple."):v(_)},[]),T=()=>{p(!0),f.current=!0,w(void 0),v(void 0);const _=N=>{w(C=>JSON.stringify(C)===JSON.stringify(N)?C:N),p(!1),f.current=!1};return n.subscribe(s,_).catch(console.error),()=>{n.unsubscribe(s,_).catch(console.error)}};u.useEffect(T,[n,s]);const P=()=>{if(!f.current&&typeof h<"u"){const _=setTimeout(()=>{n.set(s,h)},o);return()=>clearTimeout(_)}};u.useEffect(P,[h,n,s,o]);const E=()=>{if(JSON.stringify(h)===JSON.stringify(m)){const _=setTimeout(()=>{v(void 0)},500);return()=>clearTimeout(_)}};return u.useEffect(E,[h,m]),[c,(a=h??m)!==null&&a!==void 0?a:k,S]}function Le(n,s){return(i,o=0)=>b(i,n,s,o)}function Ie(n,s){return(i=0)=>{const o=u.useMemo(()=>x.store().instance,[]);return b(o,n,s,i)}}function Ue(n,s){return(i=0)=>{const o=u.useMemo(()=>x.store().application,[]);return b(o,n,s,i)}}function ze(n,s){return(i=0)=>{const o=u.useMemo(()=>x.store().device,[]);return b(o,n,s,i)}}function Me(n,s,i){return(o=0)=>{const a=u.useMemo(()=>x.store().shared(i),[i]);return b(a,n,s,o)}}function $e(n,s){return(i,o=0)=>{const a=u.useMemo(()=>x.store().shared(i),[i]);return b(a,n,s,o)}}function De(n){const s=()=>{document.documentElement.style.fontSize=`calc(1vmax * ${n})`};u.useEffect(s,[n])}function Ye(){const[n,s]=u.useState(()=>window.innerWidth/window.innerHeight),i=()=>{const o=()=>{s(window.innerWidth/window.innerHeight)};return window.addEventListener("resize",o),()=>window.removeEventListener("resize",o)};return u.useEffect(i,[]),n}function We(n,s){return{uiWidthFactor:Math.min(s,1)/n,uiHeightFactor:Math.max(s,1)/n}}exports.SettingsBox=ue;exports.SettingsButtonFrame=Ne;exports.SettingsCheckboxFrame=Ee;exports.SettingsCheckboxLabel=je;exports.SettingsColorFrame=ye;exports.SettingsContainer=le;exports.SettingsDivider=ge;exports.SettingsError=he;exports.SettingsField=fe;exports.SettingsHeading=de;exports.SettingsHint=me;exports.SettingsInputFrame=_e;exports.SettingsLabel=pe;exports.SettingsMediaSelect=Ce;exports.SettingsRadioFrame=Re;exports.SettingsRadioLabel=Te;exports.SettingsSection=Fe;exports.SettingsSelectFrame=be;exports.SettingsSliderFrame=ve;exports.SettingsSliderRuler=Se;exports.SettingsSwitchFrame=ke;exports.SettingsSwitchLabel=we;exports.SettingsTextAreaFrame=xe;exports.createUseApplicationStoreState=Ue;exports.createUseDeviceStoreState=ze;exports.createUseDynamicNamespaceStoreState=$e;exports.createUseInstanceStoreState=Ie;exports.createUseSharedStoreState=Me;exports.createUseStoreState=Le;exports.useStoreState=b;exports.useUiAspectRatio=Ye;exports.useUiResponsiveFactors=We;exports.useUiScaleToSetRem=De;
