// @telemetryos/cli/dist/commands/auth.js
import { Command } from 'commander';
import inquirer from 'inquirer';
import { apiTokenSchema, loadCliConfig, saveCliConfig } from '../services/cli-config.js';
export const authCommand = new Command('auth')
    .description('Authenticate with TelemetryOS by saving your API token')
    .option('-t, --token <string>', 'API token (skip interactive prompt)')
    .option('-f, --force', 'Replace existing token without confirmation')
    .action(handleAuthCommand);
function maskToken(token) {
    if (token.length <= 8)
        return '****';
    return `${token.slice(0, 4)}...${token.slice(-4)}`;
}
function validateToken(input) {
    const result = apiTokenSchema.safeParse(input.trim());
    if (result.success)
        return true;
    return result.error.issues[0].message;
}
async function handleAuthCommand(options) {
    let token = options.token;
    const existingCliConfig = await loadCliConfig();
    const existingToken = existingCliConfig.apiToken;
    if (existingToken && !options.force) {
        const { confirm } = await inquirer.prompt([
            {
                type: 'confirm',
                name: 'confirm',
                message: `A token already exists (${maskToken(existingToken)}). Do you want to replace it?`,
                default: false,
            },
        ]);
        if (!confirm) {
            console.log('Authentication cancelled.');
            return;
        }
    }
    if (token) {
        const validation = validateToken(token);
        if (validation !== true) {
            console.error(validation);
            process.exit(1);
        }
        token = token.trim();
    }
    else {
        console.log('In Studio, go to Settings > API Tokens to get your API token.\n');
        const answers = await inquirer.prompt([
            {
                type: 'input',
                name: 'token',
                message: 'Paste your API token:',
                validate: validateToken,
            },
        ]);
        token = answers.token.trim();
    }
    await saveCliConfig({ ...existingCliConfig, apiToken: token });
    console.log('API token saved successfully!');
}
