// @telemetryos/cli/dist/commands/init.js
import { Command } from 'commander';
import { createProject, checkDirectoryConflicts, removeConflictingFiles, } from '../services/create-project.js';
import inquirer from 'inquirer';
import path from 'path';
import { resolveProjectPathAndName, validateProjectName } from '../utils/path-utils.js';
import { ansi } from '../utils/ansi.js';
async function promptConflictResolution() {
    const { action } = await inquirer.prompt([
        {
            type: 'list',
            name: 'action',
            message: 'Target directory is not empty. Please choose how to proceed:',
            choices: [
                { name: 'Cancel operation', value: 'cancel' },
                { name: 'Remove existing files and continue', value: 'remove' },
                { name: 'Ignore files and continue', value: 'ignore' },
            ],
            default: 'cancel',
        },
    ]);
    return action;
}
export const initCommand = new Command('init')
    .description('Initializes a new telemetryOS application')
    .argument('[project-path]', 'Path to create the project (defaults to current directory)', '')
    .option('-d, --description <string>', 'The description of the application', '')
    .option('-a, --author <string>', 'The author of the application', '')
    .option('-v, --version <string>', 'The version of the application', '0.1.0')
    .option('-t, --template <string>', 'The template to use (vite-react-typescript, vite-react-typescript-web)', '')
    .option('-y, --yes', 'Skip all prompts and use defaults', false)
    .action(handleInitCommand);
async function handleInitCommand(projectPathArg, options) {
    var _a, _b, _c, _d, _e, _f;
    // Step 1: Resolve path and derive name
    const cwd = process.cwd();
    const inputPath = projectPathArg || '.';
    const { resolvedPath, derivedName } = resolveProjectPathAndName(inputPath, cwd);
    // Step 2: Validate the derived name BEFORE any destructive operations
    const validation = validateProjectName(derivedName);
    if (validation !== true) {
        console.error(`\n${ansi.red}Error:${ansi.reset} Cannot initialize project - ${validation}`);
        console.error(`\nDerived project name: ${ansi.yellow}${derivedName}${ansi.reset}`);
        console.error(`From path: ${ansi.dim}${path.relative(cwd, resolvedPath)}${ansi.reset}`);
        console.error(`\n${ansi.bold}Project names must:${ansi.reset}`);
        console.error('  • Contain only lowercase letters, numbers, and hyphens');
        console.error('  • Not start or end with a hyphen');
        console.error('  • Not start with a dot (.) or underscore (_)');
        console.error('  • Be between 1 and 214 characters');
        console.error('  • Not be a reserved name (e.g., node_modules)');
        console.error(`\n${ansi.yellow}Tip:${ansi.reset} Use a different directory name or specify a path`);
        console.error(`Example: ${ansi.cyan}tos init my-app${ansi.reset}`);
        process.exit(1);
    }
    // Step 3: Check for directory conflicts BEFORE prompting for details
    const conflicts = await checkDirectoryConflicts(resolvedPath);
    if (conflicts.length > 0) {
        console.log(`\n${ansi.yellow}⚠${ansi.reset}  Target directory contains files that could conflict:\n`);
        conflicts.slice(0, 10).forEach((file) => console.log(`  ${ansi.dim}${file}${ansi.reset}`));
        if (conflicts.length > 10) {
            console.log(`  ${ansi.dim}... and ${conflicts.length - 10} more${ansi.reset}`);
        }
        console.log();
        const action = await promptConflictResolution();
        if (action === 'cancel') {
            console.log('Operation cancelled');
            process.exit(0);
        }
        else if (action === 'remove') {
            console.log('\nRemoving existing files...');
            await removeConflictingFiles(resolvedPath, conflicts);
            console.log(`${ansi.green}✓${ansi.reset} Existing files removed\n`);
        }
        else {
            // action === 'ignore' - continue with merge
            console.log('\nContinuing with existing files (will overwrite conflicts)...\n');
        }
    }
    // Step 4: Setup variables from options
    let name = derivedName;
    let description = options.description;
    let author = options.author;
    let version = options.version;
    let template = options.template;
    let claudeCode = true;
    // Step 5: Build prompt questions (skipped with --yes)
    if (options.yes) {
        if (!name) {
            console.error('Project name is required when using --yes');
            process.exit(1);
        }
        if (!description)
            description = 'A telemetryOS application';
        if (!template)
            template = 'vite-react-typescript';
        if (!version)
            version = '0.1.0';
    }
    else {
        const questions = [];
        // Always show the derived name with option to override
        // This provides transparency and control
        questions.push({
            type: 'input',
            name: 'name',
            message: 'Project name:',
            default: derivedName,
            validate: (input) => {
                const result = validateProjectName(input);
                return result === true ? true : result;
            },
        });
        if (!description)
            questions.push({
                type: 'input',
                name: 'description',
                message: 'What is the description of your application?',
                default: 'A telemetryOS application',
            });
        if (!author)
            questions.push({
                type: 'input',
                name: 'author',
                message: 'Who is the author of your application?',
                default: '',
            });
        if (!version)
            questions.push({
                type: 'input',
                name: 'version',
                message: 'What is the version of your application?',
                default: '0.1.0',
                validate: (input) => {
                    if (/^\d+\.\d+\.\d+(-.+)?$/.test(input))
                        return true;
                    return 'Version must be in semver format (e.g. 1.0.0)';
                },
            });
        if (!template)
            questions.push({
                type: 'list',
                name: 'template',
                message: 'Which template would you like to use?',
                choices: [
                    { name: 'Vite + React + TypeScript', value: 'vite-react-typescript' },
                    { name: 'Vite + React + TypeScript + Web', value: 'vite-react-typescript-web' },
                ],
            });
        questions.push({
            type: 'confirm',
            name: 'claudeCode',
            message: 'Include Claude Code skills and settings?',
            default: true,
        });
        // Step 6: Prompt user
        const answers = await inquirer.prompt(questions);
        name = (_a = answers.name) !== null && _a !== void 0 ? _a : name;
        version = (_b = answers.version) !== null && _b !== void 0 ? _b : version;
        description = (_c = answers.description) !== null && _c !== void 0 ? _c : description;
        author = (_d = answers.author) !== null && _d !== void 0 ? _d : author;
        template = (_e = answers.template) !== null && _e !== void 0 ? _e : template;
        claudeCode = (_f = answers.claudeCode) !== null && _f !== void 0 ? _f : claudeCode;
    }
    // Step 7: Generate application
    await createProject({
        name,
        description,
        author,
        version,
        template,
        claudeCode,
        projectPath: resolvedPath,
        progressFn: (createdFilePath) => {
            console.log(`.${path.sep}${path.relative(cwd, createdFilePath)}`);
        },
    });
}
