// @telemetryos/cli/dist/commands/publish.js
import { Command } from 'commander';
import inquirer from 'inquirer';
import path from 'path';
import { ApiClient } from '../services/api-client.js';
import { loadProjectConfig } from '../services/project-config.js';
import { createArchive } from '../services/archiver.js';
import { pollBuild } from '../services/build-poller.js';
import { ansi } from '../utils/ansi.js';
// Default values for new application builds
const APP_BUILD_DEFAULTS = {
    baseImage: 'node:20',
    buildScript: 'npm install\nnpm run build',
    buildOutput: '/dist',
    workingPath: '/',
};
export const publishCommand = new Command('publish')
    .description('Publish an application to TelemetryOS')
    .argument('[project-path]', 'Path to the project directory. Defaults to current working directory', process.cwd())
    .option('-i, --interactive', 'Interactively prompt for all build configuration')
    .option('--name <name>', 'Application name (overrides config)')
    .option('--base-image <image>', 'Docker base image for build')
    .option('--build-script <script>', 'Build script commands')
    .option('--build-output <path>', 'Build output directory')
    .option('--working-path <path>', 'Working directory in container')
    .action(handlePublishCommand);
function getSpecifiedBuildOptions(options) {
    const specified = [];
    if (options.baseImage !== undefined)
        specified.push('--base-image');
    if (options.buildScript !== undefined)
        specified.push('--build-script');
    if (options.buildOutput !== undefined)
        specified.push('--build-output');
    if (options.workingPath !== undefined)
        specified.push('--working-path');
    return specified;
}
async function handlePublishCommand(projectPath, options, callbacks) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q;
    projectPath = path.resolve(process.cwd(), projectPath);
    const logMessage = (callbacks === null || callbacks === void 0 ? void 0 : callbacks.onLog) || ((msg) => console.log(msg));
    try {
        // Step 1: Load project config
        (_a = callbacks === null || callbacks === void 0 ? void 0 : callbacks.onStateChange) === null || _a === void 0 ? void 0 : _a.call(callbacks, 'loading');
        logMessage(`\n${ansi.cyan}Loading project configuration...${ansi.reset}`);
        const config = await loadProjectConfig(projectPath);
        // Step 2: Resolve build configuration
        let buildConfig = {
            name: options.name || config.name || '',
            baseImage: options.baseImage,
            buildScript: options.buildScript,
            buildOutput: options.buildOutput,
            workingPath: options.workingPath,
        };
        if (options.interactive) {
            // Interactive mode: prompt for all fields
            buildConfig = await promptBuildConfig(buildConfig);
        }
        else if (!buildConfig.name) {
            // Non-interactive mode but no name: prompt just for name
            const answers = await inquirer.prompt([
                {
                    type: 'input',
                    name: 'name',
                    message: 'Application name:',
                    validate: (input) => (input.trim() ? true : 'Application name is required'),
                },
            ]);
            buildConfig.name = answers.name.trim();
        }
        if (!buildConfig.name) {
            const error = new Error('Application name is required');
            logMessage(`${ansi.red}Error: Application name is required${ansi.reset}`);
            (_b = callbacks === null || callbacks === void 0 ? void 0 : callbacks.onError) === null || _b === void 0 ? void 0 : _b.call(callbacks, error);
            if (!callbacks)
                process.exit(1);
            return;
        }
        logMessage(`  Project: ${ansi.bold}${buildConfig.name}${ansi.reset}`);
        if (config.version) {
            logMessage(`  Version: ${config.version}`);
        }
        // Step 3: Initialize API client
        (_c = callbacks === null || callbacks === void 0 ? void 0 : callbacks.onStateChange) === null || _c === void 0 ? void 0 : _c.call(callbacks, 'authenticating');
        logMessage(`\n${ansi.cyan}Authenticating...${ansi.reset}`);
        const apiClient = await ApiClient.create();
        logMessage(`  ${ansi.green}Authenticated${ansi.reset}`);
        // Step 4: Create archive
        (_d = callbacks === null || callbacks === void 0 ? void 0 : callbacks.onStateChange) === null || _d === void 0 ? void 0 : _d.call(callbacks, 'archiving');
        logMessage(`\n${ansi.cyan}Archiving project...${ansi.reset}`);
        const { buffer, filename } = await createArchive(projectPath, (msg) => {
            logMessage(`  ${ansi.dim}${msg}${ansi.reset}`);
        });
        // Step 5: Find or create application
        logMessage(`\n${ansi.cyan}Checking for existing application...${ansi.reset}`);
        const appsRes = await apiClient.get('/application');
        const apps = (await appsRes.json());
        let app = apps.find((a) => a.title === buildConfig.name) || null;
        if (app) {
            logMessage(`  Found existing application: ${ansi.dim}${app.id}${ansi.reset}`);
            const specifiedOptions = getSpecifiedBuildOptions(options);
            if (specifiedOptions.length > 0) {
                logMessage(`  ${ansi.yellow}Warning: ${specifiedOptions.join(', ')} ignored for existing application${ansi.reset}`);
            }
        }
        else {
            logMessage(`  Creating new application...`);
            const createRequest = {
                kind: 'uploaded',
                title: buildConfig.name,
                baseImage: (_e = buildConfig.baseImage) !== null && _e !== void 0 ? _e : APP_BUILD_DEFAULTS.baseImage,
                baseImageRegistryAuth: '',
                buildWorkingPath: (_f = buildConfig.workingPath) !== null && _f !== void 0 ? _f : APP_BUILD_DEFAULTS.workingPath,
                buildScript: (_g = buildConfig.buildScript) !== null && _g !== void 0 ? _g : APP_BUILD_DEFAULTS.buildScript,
                buildOutputPath: (_h = buildConfig.buildOutput) !== null && _h !== void 0 ? _h : APP_BUILD_DEFAULTS.buildOutput,
            };
            const createRes = await apiClient.post('/application', { body: createRequest });
            app = (await createRes.json());
            logMessage(`  ${ansi.green}Created application:${ansi.reset} ${ansi.dim}${app.id}${ansi.reset}`);
        }
        // Step 6: Upload archive
        (_j = callbacks === null || callbacks === void 0 ? void 0 : callbacks.onStateChange) === null || _j === void 0 ? void 0 : _j.call(callbacks, 'uploading');
        logMessage(`\n${ansi.cyan}Uploading archive...${ansi.reset}`);
        const formData = new FormData();
        const arrayBuffer = buffer.buffer.slice(buffer.byteOffset, buffer.byteOffset + buffer.byteLength);
        const blob = new Blob([arrayBuffer], { type: 'application/gzip' });
        formData.append('file', blob, filename);
        await apiClient.post(`/application/${app.id}/archive`, { body: formData });
        logMessage(`  ${ansi.green}Upload complete${ansi.reset}`);
        // Step 7: Poll for build status and stream logs
        (_k = callbacks === null || callbacks === void 0 ? void 0 : callbacks.onStateChange) === null || _k === void 0 ? void 0 : _k.call(callbacks, 'building');
        logMessage(`\n${ansi.cyan}Building...${ansi.reset}`);
        logMessage(`${ansi.white}Build Logs${ansi.reset}`);
        const build = await pollBuild(apiClient, app.id, {
            onLog: (line) => {
                logMessage(`  ${ansi.dim}${line}${ansi.reset}`);
            },
            onStateChange: () => { },
            onComplete: () => { },
            onError: (error) => {
                var _a;
                logMessage(`${ansi.red}Polling error: ${error.message}${ansi.reset}`);
                (_a = callbacks === null || callbacks === void 0 ? void 0 : callbacks.onError) === null || _a === void 0 ? void 0 : _a.call(callbacks, error);
            },
        });
        // Step 8: Report result
        if (build.state === 'success') {
            (_l = callbacks === null || callbacks === void 0 ? void 0 : callbacks.onStateChange) === null || _l === void 0 ? void 0 : _l.call(callbacks, 'success');
            logMessage(`\n${ansi.green}${ansi.bold}Build successful!${ansi.reset}`);
            logMessage(`\n  Application: ${buildConfig.name}`);
            logMessage(`  Build: #${build.index}`);
            let duration = '';
            if (build.finishedAt && build.startedAt) {
                duration = calculateDuration(build.startedAt, build.finishedAt);
                logMessage(`  Duration: ${duration}`);
            }
            logMessage('');
            (_m = callbacks === null || callbacks === void 0 ? void 0 : callbacks.onComplete) === null || _m === void 0 ? void 0 : _m.call(callbacks, {
                success: true,
                buildIndex: build.index,
                duration,
            });
        }
        else {
            (_o = callbacks === null || callbacks === void 0 ? void 0 : callbacks.onStateChange) === null || _o === void 0 ? void 0 : _o.call(callbacks, 'failure');
            logMessage(`\n${ansi.red}${ansi.bold}Build failed${ansi.reset}`);
            logMessage(`\n  State: ${build.state}`);
            logMessage('');
            const error = new Error(`Build failed with state: ${build.state}`);
            (_p = callbacks === null || callbacks === void 0 ? void 0 : callbacks.onError) === null || _p === void 0 ? void 0 : _p.call(callbacks, error);
            if (!callbacks) {
                process.exit(1);
            }
        }
    }
    catch (error) {
        const err = error;
        logMessage(`\n${ansi.red}Error: ${err.message}${ansi.reset}\n`);
        (_q = callbacks === null || callbacks === void 0 ? void 0 : callbacks.onError) === null || _q === void 0 ? void 0 : _q.call(callbacks, err);
        if (!callbacks) {
            process.exit(1);
        }
    }
}
// Export for use by server
export { handlePublishCommand };
async function promptBuildConfig(defaults) {
    var _a, _b, _c, _d;
    const answers = await inquirer.prompt([
        {
            type: 'input',
            name: 'name',
            message: 'Application name:',
            default: defaults.name || undefined,
            validate: (input) => (input.trim() ? true : 'Application name is required'),
        },
        {
            type: 'input',
            name: 'baseImage',
            message: 'Base image:',
            default: (_a = defaults.baseImage) !== null && _a !== void 0 ? _a : APP_BUILD_DEFAULTS.baseImage,
        },
        {
            type: 'input',
            name: 'buildScript',
            message: 'Build script:',
            default: (_b = defaults.buildScript) !== null && _b !== void 0 ? _b : APP_BUILD_DEFAULTS.buildScript,
        },
        {
            type: 'input',
            name: 'buildOutput',
            message: 'Build output path:',
            default: (_c = defaults.buildOutput) !== null && _c !== void 0 ? _c : APP_BUILD_DEFAULTS.buildOutput,
        },
        {
            type: 'input',
            name: 'workingPath',
            message: 'Working path:',
            default: (_d = defaults.workingPath) !== null && _d !== void 0 ? _d : APP_BUILD_DEFAULTS.workingPath,
        },
    ]);
    return {
        name: answers.name.trim(),
        baseImage: answers.baseImage.trim(),
        buildScript: answers.buildScript.trim(),
        buildOutput: answers.buildOutput.trim(),
        workingPath: answers.workingPath.trim(),
    };
}
function calculateDuration(start, end) {
    const ms = new Date(end).getTime() - new Date(start).getTime();
    const seconds = Math.floor(ms / 1000);
    if (seconds < 60)
        return `${seconds}s`;
    const minutes = Math.floor(seconds / 60);
    const remainingSeconds = seconds % 60;
    return `${minutes}m ${remainingSeconds}s`;
}
